File: android/app/src/main/java/com/kreativekoala/summaryai/data/api/coaching/CoachingModels.kt
package com.kreativekoala.summaryai.data.api.coaching

import com.google.gson.annotations.SerializedName

/**
 * Wire types for the `/api/coaching/...` endpoints. See backend's
 * `coaching.ts` + iOS `CoachingModels.swift` for the parallel definitions.
 */

enum class CoachingPersonaKey(val key: String, val displayName: String) {
    SalesDiscovery("sales_discovery", "Sales Discovery");

    companion object {
        fun fromKey(key: String?): CoachingPersonaKey =
            entries.firstOrNull { it.key == key } ?: SalesDiscovery
    }
}

data class CoachingCredits(
    val balance: Int,
    @SerializedName("by_source") val bySource: Map<String, Int>?
) {
    val hasCredits: Boolean get() = balance > 0
}

data class CoachingClaimFreeResponse(
    val granted: Boolean,
    val balance: Int,
    @SerializedName("free_tier_credits") val freeTierCredits: Int
)

data class CoachingDebugGrantResponse(
    val granted: Int,
    val balance: Int
)

data class CoachingStartSessionRequest(
    @SerializedName("recording_id") val recordingId: String,
    val persona: String
)

data class CoachingStartSessionResponse(
    @SerializedName("session_id") val sessionId: String,
    val persona: String
)

data class CoachingInsight(
    val id: String,
    @SerializedName("session_id") val sessionId: String,
    val type: String,         // "question" | "objection" | "signal" | "gap"
    val text: String,
    val urgency: String,      // "now" | "soon" | "before-end"
    @SerializedName("transcript_offset_seconds") val transcriptOffsetSeconds: Int?,
    @SerializedName("emitted_at") val emittedAt: String
)

data class CoachingInsightsResponse(
    val session: CoachingSessionMeta?,
    val insights: List<CoachingInsight>
)

data class CoachingSessionMeta(
    @SerializedName("user_id") val userId: String?,
    val persona: String?,
    @SerializedName("started_at") val startedAt: String?,
    @SerializedName("ended_at") val endedAt: String?,
    @SerializedName("insight_count") val insightCount: Int?,
    val refunded: Boolean?
)

data class CoachingEndResponse(
    val ended: Boolean,
    val refunded: Boolean
)

data class CoachingGooglePlayVerifyRequest(
    @SerializedName("purchase_token") val purchaseToken: String,
    @SerializedName("product_id") val productId: String
)

data class CoachingGooglePlayVerifyResponse(
    val granted: Boolean,
    @SerializedName("already_existed") val alreadyExisted: Boolean,
    val balance: Int,
    @SerializedName("expires_at") val expiresAt: String?
)

File: ios/SummaryAI/Sources/Models/CoachingModels.swift
import Foundation

/// Mirror of the backend's `/api/coaching/*` response shapes. See
/// REALTIME_COACHING.md for the design.

enum CoachingPersonaKey: String, Codable, CaseIterable {
    case salesDiscovery = "sales_discovery"

    var displayName: String {
        switch self {
        case .salesDiscovery: return "Sales Discovery"
        }
    }
}

struct CoachingCredits: Codable {
    let balance: Int
    let bySource: [String: Int]

    enum CodingKeys: String, CodingKey {
        case balance
        case bySource = "by_source"
    }

    var hasCredits: Bool { balance > 0 }
}

struct CoachingClaimFreeResponse: Codable {
    let granted: Bool
    let balance: Int
    let freeTierCredits: Int

    enum CodingKeys: String, CodingKey {
        case granted, balance
        case freeTierCredits = "free_tier_credits"
    }
}

struct CoachingSessionStart: Codable {
    let sessionId: String
    let persona: String

    enum CodingKeys: String, CodingKey {
        case sessionId = "session_id"
        case persona
    }
}

struct CoachingInsight: Codable, Identifiable, Equatable {
    let id: String
    let sessionId: String
    let type: InsightType
    let text: String
    let urgency: Urgency
    let transcriptOffsetSeconds: Int?
    let emittedAt: Date

    enum InsightType: String, Codable {
        case question, objection, signal, gap
    }
    enum Urgency: String, Codable {
        case now, soon
        case beforeEnd = "before-end"

        var sortKey: Int {
            switch self { case .now: return 0; case .soon: return 1; case .beforeEnd: return 2 }
        }
    }

    enum CodingKeys: String, CodingKey {
        case id
        case sessionId = "session_id"
        case type, text, urgency
        case transcriptOffsetSeconds = "transcript_offset_seconds"
        case emittedAt = "emitted_at"
    }
}

struct CoachingEndResponse: Codable {
    let ended: Bool
    let refunded: Bool
}
